core/src/main/java/com/graphhopper/routing/weighting/custom/SpeedCustomConfig.java

```diff
@@ -0,0 +1,132 @@
+                }
+            } else if (value instanceof Map) {
+                EnumEncodedValue enumEncodedValue = getEV(lookup, "max_speed", key, EnumEncodedValue.class);
+                Class<? extends Enum> enumClass = factory.findValues(key);
+                double[] values = Helper.createEnumToDoubleArray("max_speed." + key, maxSpeed, 0, maxSpeed,
+                        enumClass, (Map<String, Object>) value);
+                maxSpeedList.add(new EnumToValueEntry(enumEncodedValue, values));
+            } else {
+                throw new IllegalArgumentException("Type " + value.getClass() + " is not supported for 'max_speed'");
+            }
+        }
+
+        // use speed_factor to reduce the estimated speed value under the specified conditions
+        for (Map.Entry<String, Object> entry : customModel.getSpeedFactor().entrySet()) {
+            String key = entry.getKey();
+            Object value = entry.getValue();
+
+            if (value instanceof Number) {
+                if (key.startsWith(GeoToValueEntry.key(""))) {
+                    Geometry geometry = GeoToValueEntry.pickGeometry(customModel, key);
+                    speedFactorList.add(new GeoToValueEntry(new PreparedGeometryFactory().create(geometry), ((Number) value).doubleValue(), 1));
+                } else {
+                    BooleanEncodedValue encodedValue = getEV(lookup, "speed_factor", key, BooleanEncodedValue.class);
+                    speedFactorList.add(new BooleanToValueEntry(encodedValue, ((Number) value).doubleValue(), 1));
+                }
+            } else if (value instanceof Map) {
+                EnumEncodedValue enumEncodedValue = getEV(lookup, "speed_factor", key, EnumEncodedValue.class);
+                Class<? extends Enum> enumClass = factory.findValues(key);
+                double[] values = Helper.createEnumToDoubleArray("speed_factor." + key, 1, 0, 1,
+                        enumClass, (Map<String, Object>) value);
+                speedFactorList.add(new EnumToValueEntry(enumEncodedValue, values));
+            } else {
+                throw new IllegalArgumentException("Type " + value.getClass() + " is not supported for 'speed_factor'");
+            }
+        }
+    }
+
+    public double getMaxSpeed() {
+        return maxSpeed;
+    }
+
+    /**
+     * @return speed in km/h
+     */
+    public double calcSpeed(EdgeIteratorState edge, boolean reverse) {
+        double speed = reverse ? edge.getReverse(avgSpeedEnc) : edge.get(avgSpeedEnc);
+        if (Double.isInfinite(speed) || Double.isNaN(speed) || speed < 0)
+            throw new IllegalStateException("Invalid estimated speed " + speed);
+
+        for (int i = 0; i < speedFactorList.size(); i++) {
+            EdgeToValueEntry entry = speedFactorList.get(i);
+            double factorValue = entry.getValue(edge, reverse);
+            speed *= factorValue;
+            if (speed <= 0)
+                break;
```

Comment:
Similar question as in `PriorityCustomConfig`: Why do we keep going if `entry.getValue()` returns zero or something negative?